NetKet/Hamiltonian/MatrixWrapper/matrix_wrapper.hpp

```diff
@@ -0,0 +1,40 @@
+#ifndef MATRIX_WRAPPER_HPP
+#define MATRIX_WRAPPER_HPP
+
+#include <memory>
+
+#include "Graph/graph.hpp"
+#include "Hamiltonian/hamiltonian.hpp"
+#include "Utils/json_helper.hpp"
+
+#include "dense_matrix_wrapper.hpp"
+#include "direct_matrix_wrapper.hpp"
+#include "sparse_matrix_wrapper.hpp"
+
+namespace netket {
+
+template<class Wrapped>
+std::unique_ptr<AbstractMatrixWrapper<Wrapped>>
+ConstructMatrixWrapper(const json &pars, const Wrapped& wrapped)
+{
+    using WrapperPtr = std::unique_ptr<AbstractMatrixWrapper<Wrapped>>;
+
+    std::string wrapper_name = FieldOrDefaultVal<json, std::string>(pars, "MatrixWrapper", "Sparse");
+    if(wrapper_name == "Sparse")
+    {
+        return WrapperPtr(new SparseMatrixWrapper<Wrapped>(wrapped));
```

Comment:
This is *not* exception safe. You're copying `wrapped` (which is a bad idea in the first place) whose copy constructor may throw an exception. Now, if an exception is thrown during construction, the newly allocated pointer will never be freed, i.e. you're leaking memory.

Reply:
The constructor of `SparseMatrixWrapper` takes `const Wrapped&` so there should not be a copy involved. Or am I missing something?

Reply:
There sure is. You're storing a matrix, not a reference to it. By passing `Wrapped const&` you're invoking a [copy constructor](http://en.cppreference.com/w/cpp/language/copy_constructor).